ddtrace/propagation/http.py

```diff
@@ -483,10 +489,201 @@ def _extract(headers):
         return None
 
 
+class _TraceContext:
+    """Helper class to inject/extract W3C Trace Context
+    https://www.w3.org/TR/trace-context/
+    Overview:
+      - ``traceparent`` header describes the position of the incoming request in its
+        trace graph in a portable, fixed-length format. Its design focuses on
+        fast parsing. Every tracing tool MUST properly set traceparent even when
+        it only relies on vendor-specific information in tracestate
+      - ``tracestate`` header extends traceparent with vendor-specific data represented
+        by a set of name/value pairs. Storing information in tracestate is
+        optional.
+
+    The format for ``traceparent`` is::
+      HEXDIGLC        = DIGIT / "a" / "b" / "c" / "d" / "e" / "f"
+      value           = version "-" version-format
+      version         = 2HEXDIGLC
+      version-format  = trace-id "-" parent-id "-" trace-flags
+      trace-id        = 32HEXDIGLC
+      parent-id       = 16HEXDIGLC
+      trace-flags     = 2HEXDIGLC
+
+    Example value of HTTP ``traceparent`` header::
+        value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
+        base16(version) = 00
+        base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
+        base16(parent-id) = 00f067aa0ba902b7
+        base16(trace-flags) = 01  // sampled
+
+    The format for ``tracestate`` is key value pairs with each entry limited to 256 characters.
+    An example of the ``dd`` list member we would add is::
+    "dd=s:2;o:rum;t.dm:-4;t.usr.id:baz64"
+
+    Implementation details:
+      - Datadog Trace and Span IDs are 64-bit unsigned integers.
+      - The W3C Trace Context Trace ID is a 16-byte hexadecimal string.
+      - If the incoming traceparent was valid we DO NOT update the trace-id.
+        Otherwise, the value is set to the hex-encoded value of the trace-id.
+        If the trace-id is a 64-bit value (i.e. a Datadog trace-id),
+        then the upper half of the hex-encoded value will be all zeroes.
+
+      - The tracestate header will have one list member added to it, ``dd``, which contains
+        values that would be in x-datadog-tags as well as those needed for propagation information.
+        The keys to the ``dd`` values have been shortened as follows to save space:
+        ``sampling_priority`` = ``s``
+        ``origin`` = ``o``
+        ``_dd.p.`` prefix = ``t.``
+    """
+
+    @staticmethod
+    def _is_valid_datadog_trace_tag_key(key):
+        return key.startswith("t.")
+
+    @staticmethod
+    def _get_traceparent_values(headers):
+        # type: (Dict[str, str]) -> Optional[Tuple[str,int,int,int]]
+
+        tp = _extract_header_value(_POSSIBLE_HTTP_HEADER_TRACEPARENT, headers)
+        if tp is None:
+            return None
+
+        try:
+            version, trace_id_hex, span_id_hex, trace_flags_hex = tp.split("-")
+            # check version is a valid hexadecimal
+            int(version, 16)
```

Comment:
Do we need to do this check? Doesn't seem like it gains us much and if we don't have a supported version we still continue anyway (see below)

Reply:
I'm not sure, I guess in my mind there's a difference between version being a valid hexadecimal that we haven't added support for yet, and version being an invalid combination of characters, in which case we should return `None`. Do you still think we should remove it though?

Reply:
🤷‍♂️ I don't feel strongly one way or the other. I do like being louder about invalid data rather than pretending it's a valid version and that it isn't supported. Louder and more explicit would help the user more if something were to go wrong. Maybe just worth updating the comment to say this more explicitly.

Reply:
I'll update the comment!